test/integration/dualstack/dualstack_endpoints_test.go

```diff
@@ -0,0 +1,294 @@
+			if err != nil {
+				t.Fatalf("Failed to create pod %s: %v", pod.Name, err)
+			}
+
+			// Set pod IPs
+			podIPbyFamily := map[v1.IPFamily]string{v1.IPv4Protocol: "1.1.1.1", v1.IPv6Protocol: "2001:db2::65"}
+			createdPod.Status = v1.PodStatus{
+				Phase:  v1.PodRunning,
+				PodIPs: []v1.PodIP{{IP: podIPbyFamily[v1.IPv4Protocol]}, {IP: podIPbyFamily[v1.IPv6Protocol]}},
+			}
+			_, err = client.CoreV1().Pods(ns.Name).UpdateStatus(context.TODO(), createdPod, metav1.UpdateOptions{})
+			if err != nil {
+				t.Fatalf("Failed to update status of pod %s: %v", pod.Name, err)
+			}
+
+			svc := &v1.Service{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      fmt.Sprintf("svc-test-%d", i), // use different services for each test
+					Namespace: ns.Name,
+					Labels:    labelMap(),
+				},
+				Spec: v1.ServiceSpec{
+					Type:           v1.ServiceTypeClusterIP,
+					IPFamilies:     tc.ipFamilies,
+					IPFamilyPolicy: &tc.ipFamilyPolicy,
+					Selector:       labelMap(),
+					Ports: []v1.ServicePort{
+						{
+							Name:       fmt.Sprintf("port-test-%d", i),
+							Port:       443,
+							TargetPort: intstr.IntOrString{IntVal: 443},
+							Protocol:   "TCP",
+						},
+					},
+				},
+			}
+
+			// create a service
+			_, err = client.CoreV1().Services(ns.Name).Create(context.TODO(), svc, metav1.CreateOptions{})
+			if err != nil {
+				t.Fatalf("error creating service: %v", err)
+			}
+
+			// wait until endpoints are created
+			// legacy endpoints are not dual stack
+			// and use the address of the first IP family
+			if err := wait.PollImmediate(1*time.Second, wait.ForeverTestTimeout, func() (bool, error) {
+				e, err := client.CoreV1().Endpoints(ns.Name).Get(context.TODO(), svc.Name, metav1.GetOptions{})
+				if err != nil {
+					t.Logf("error fetching endpoints: %v", err)
+					return false, nil
+				}
+				// check if the endpoint addresses match the pod IP of the first IPFamily of the service
+				// since this is an integration test PodIPs are not "ready"
+				if len(e.Subsets) > 0 && len(e.Subsets[0].NotReadyAddresses) > 0 {
+					if e.Subsets[0].NotReadyAddresses[0].IP == podIPbyFamily[tc.ipFamilies[0]] {
+						return true, nil
+					}
+					t.Logf("endpoint address %s does not match PodIP %s ", e.Subsets[0].Addresses[0].IP, podIPbyFamily[tc.ipFamilies[0]])
+				}
+				t.Logf("endpoint does not contain addresses: %s", e.Name)
+				return false, nil
+			}); err != nil {
+				t.Fatalf("endpoints not found: %v", err)
+			}
+
+			// wait until the endpoint slices are created
+			err = wait.PollImmediate(1*time.Second, wait.ForeverTestTimeout, func() (bool, error) {
+				lSelector := discovery.LabelServiceName + "=" + svc.Name
+				esList, err := client.DiscoveryV1beta1().EndpointSlices(ns.Name).List(context.TODO(), metav1.ListOptions{LabelSelector: lSelector})
+				if err != nil {
+					t.Logf("error listing EndpointSlices: %v", err)
+					return false, err
```

Comment:
the above poll for `client.CoreV1().Endpoints(ns.Name)` seems to return `false, nil` here.
should an error listing the DiscoveryV1beta1 slices exit the poll instead of retrying?

Reply:
good catch